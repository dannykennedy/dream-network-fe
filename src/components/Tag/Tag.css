/* TAG */
.tag {
    border-radius: 5px;
    width: auto;
    color: white;
    display: inline-block;
    padding: 4px 7px;
    margin-right: 10px;
    margin-bottom: 5px;
    cursor: pointer;
    font-family: "Open Sans", sans-serif;
    font-weight: 500;
    -webkit-font-smoothing: antialiased;
}

.tag-wrapper {
    display: flex;
    flex-direction: row;
}

.tag-options {
    display: flex;
    flex-direction: row;
}

.tag-icon {
}

.tag-icon-dropdown {
    padding: 3px 5px;
}

.tag-edit-input {
    margin-left: 5px;
}

.date {
    background-color: rgb(230, 133, 133);
}
.date:hover {
    background-color: rgb(222, 93, 93);
    text-decoration: none;
    color: white;
}
.person {
    background-color: rgb(246, 173, 97);
}
.person:hover {
    background-color: rgba(244, 152, 58);
    text-decoration: none;
    color: white;
}
.location {
    background-color: rgb(154, 200, 134);
}
.location:hover {
    background-color: rgb(125, 184, 99);
    text-decoration: none;
    color: white;
}
.other {
    background-color: rgb(121, 174, 245);
}
.other:hover {
    background-color: rgb(82, 151, 242);
    text-decoration: none;
    color: white;
}
.none {
    background-color: rgba(148, 147, 185, 0.7);
}
.none:hover {
    background-color: rgba(148, 147, 185, 0.9);
    text-decoration: none;
    color: white;
}

.tag-name {
    margin-left: 6px;
    margin-right: 6px;
    padding: 3px;
}

/* INPUT */
.tag-wrapper input {
    font-family: "Open Sans", sans-serif;
    font-weight: 500;
    -webkit-font-smoothing: antialiased;
    font-size: 1em;
    margin: 0 4px 0 1px;
    color: white;
    display: inline-block;
    background-color: rgba(0, 0, 0, 0.2);
    border: none;
    border-radius: 0.25rem;
    padding: 2px 5px;
    border-radius: 5px;
    height: 25px;
}

.tag-wrapper input:focus {
    outline: none;
}

.tag-wrapper *::placeholder {
    color: white;
    opacity: 0.7;
}

/* END TAG */
