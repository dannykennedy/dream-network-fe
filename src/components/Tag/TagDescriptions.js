// Principles 1:
// Both the tag description and the tag type should be in an "is-a" relationship with the tag type
// E.g.
// an author is a person, and Kelly Bulkeley is a person
// Date posted is a date, and 14 May 2015 is a date
// Principles 2
export const TagDescriptions = {
    PERSON: ["Author", "Subject", "(None)"],
    DATE: ["Posted on", "(None)"],
    OTHER: ["Subject", "(None)"],
};
