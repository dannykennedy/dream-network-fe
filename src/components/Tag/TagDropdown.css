.tag-dropdown {
    border-radius: 3px;
}

.tag-dropbtn {
    height: 100%;
    width: 30px;
    color: white;
    font-size: 1em;
    cursor: pointer;
    border-radius: 3px;
}

.tag-dropbtn:hover,
.tag-dropbtn:focus {
    color: white;
    background-color: transparent;
}

.tag-dropdown-content {
    box-sizing: border-box;
    position: absolute;
    overflow: auto;
    box-shadow: 0px 8px 16px 0px rgba(0, 0, 0, 0.2);
    z-index: 1;
    width: 30px;
    background-color: white;
    padding: 0px 2px 2px 2px;
    border-bottom-left-radius: 3px;
    border-bottom-right-radius: 3px;
}

.tag-dropdown-content button {
    color: white;
    text-decoration: none;
    font-size: inherit;
    display: block;
    width: 100%;
    border: none;
    width: 26px;
    height: 26px;
    border-radius: 3px;
    margin-top: 2px;
}

@media (max-width: 800px) {
    /* Make sure dropdown doesn't overflow the screen */
    .tag-dropdown-content {
        transform: translate(-88%);
    }
}
