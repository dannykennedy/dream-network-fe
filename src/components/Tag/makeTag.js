export const makeTag = (
    tagId,
    tagName,
    tagType,
    itemId,
    isNewTag,
    tagDescription
) => {
    return {
        tagId: tagId,
        tagName: tagName,
        tagType: tagType,
        itemId: itemId,
        isNewTag: isNewTag,
        tagDescription: tagDescription,
    };
};
