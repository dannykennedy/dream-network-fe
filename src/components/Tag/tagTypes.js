export const tagTypes = {
    PERSON: "PERSON",
    LOCATION: "LOCATION",
    OTHER: "OTHER",
    DATE: "DATE",
    DOCUMENT: "DOCUMENT",
};

export const mapEntitiesToTypes = {
    PERSON: tagTypes.PERSON,
    LOCATION: tagTypes.LOCATION,
    DATE: tagTypes.DATE,
    OTHER: tagTypes.OTHER,
    ORGANIZATION: tagTypes.LOCATION,
    NUMBER: tagTypes.OTHER,
    EVENT: tagTypes.OTHER,
    WORK_OF_ART: tagTypes.OTHER,
    CONSUMER_GOOD: tagTypes.OTHER,
    DOCUMENT: tagTypes.DOCUMENT,
    NONE: "NONE",
};
