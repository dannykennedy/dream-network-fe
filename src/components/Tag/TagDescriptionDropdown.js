import React, { useEffect, useState, useRef } from "react";
import FontAwesome from "react-fontawesome";
import "./TagDescriptionDropdown.css";
import { connect } from "react-redux";
import { tagColors, hoveredTagColors } from "../../theme/Theme";
import { setTagDescription as _setTagDescription } from "../../ducks/posts";
import { TagDescriptions } from "./TagDescriptions";

// https://medium.com/@pitipatdop/little-neat-trick-to-capture-click-outside-with-react-hook-ba77c37c7e82
const TagDescriptionDropdown = ({
    postId,
    tagId,
    tagType,
    setTagDescription,
    tagDescription,
}) => {
    const node = useRef();

    const [dropdownIsOpen, setDropdownIsOpen] = useState(false);

    const handleClick = e => {
        if (node.current.contains(e.target)) {
            return; // inside click
        }
        setDropdownIsOpen(false); // outside click
    };

    useEffect(() => {
        document.addEventListener("mousedown", handleClick);

        return () => {
            document.removeEventListener("mousedown", handleClick);
        };
    }, []);

    return (
        <div
            ref={node}
            className="tag-dropdown"
            style={{
                backgroundColor: "white",
            }}
        >
            <button
                style={{
                    backgroundColor: tagColors[tagType],
                    border: `1px solid ${hoveredTagColors[tagType]}`,
                }}
                className="tag-description-dropbtn"
                onClick={e => setDropdownIsOpen(!dropdownIsOpen)}
            >
                {tagDescription && (
                    <span className="tag-description">{tagDescription}</span>
                )}
                <FontAwesome name="caret-down" />
            </button>
            {dropdownIsOpen && (
                <div
                    className="tag-description-dropdown-content"
                    style={{ backgroundColor: tagColors[tagType] }}
                >
                    {TagDescriptions[tagType]
                        .filter(descOption => descOption !== tagDescription)
                        .map((descOption, i) => {
                            return (
                                <button
                                    className="tag-description-dropdown-button"
                                    key={i}
                                    onClick={() => {
                                        console.log(descOption);
                                        if (descOption === "(None)") {
                                            descOption = null;
                                        }
                                        setTagDescription(
                                            descOption,
                                            tagId,
                                            postId
                                        );

                                        setDropdownIsOpen(false);
                                    }}
                                >
                                    <div>
                                        <span>{descOption}</span>
                                    </div>
                                </button>
                            );
                        })}
                </div>
            )}
        </div>
    );
};

const mapStateToProps = state => {
    return state;
};

const mapDispatchToProps = { setTagDescription: _setTagDescription };

export default connect(
    mapStateToProps,
    mapDispatchToProps
)(TagDescriptionDropdown);
