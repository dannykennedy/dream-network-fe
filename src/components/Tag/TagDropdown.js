import React, { useEffect, useState, useRef } from "react";
import FontAwesome from "react-fontawesome";
import "./TagDropdown.css";
import { icons } from "../../theme/icons";
import { tagColors, hoveredTagColors } from "../../theme/Theme";
import { tagTypes } from "../../config/tagConfig";

// https://medium.com/@pitipatdop/little-neat-trick-to-capture-click-outside-with-react-hook-ba77c37c7e82
const TagDropdown = ({ itemId, tagId, tagType, onSetTagType }) => {
    const node = useRef();

    const [dropdownIsOpen, setDropdownIsOpen] = useState(false);

    const handleClick = e => {
        if (node.current.contains(e.target)) {
            return; // inside click
        }
        setDropdownIsOpen(false); // outside click
    };

    useEffect(() => {
        document.addEventListener("mousedown", handleClick);

        return () => {
            document.removeEventListener("mousedown", handleClick);
        };
    }, []);

    return (
        <div
            ref={node}
            className="tag-dropdown"
            style={{
                backgroundColor: "white",
            }}
        >
            <button
                style={{
                    backgroundColor: tagColors[tagType],
                    border: `1px solid ${hoveredTagColors[tagType]}`,
                }}
                className="tag-dropbtn"
                onClick={e => setDropdownIsOpen(!dropdownIsOpen)}
            >
                <FontAwesome name={icons[tagType]} />
            </button>
            {dropdownIsOpen && (
                <div
                    className="tag-dropdown-content"
                    style={{ backgroundColor: tagColors[tagType] }}
                >
                    {Object.values(tagTypes)
                        .filter(type => type !== tagType) // Don't put current tag type in dropdown
                        .map((typeOption, i) => {
                            return (
                                <button
                                    className="tag-dropdown-button"
                                    key={i}
                                    style={{
                                        backgroundColor: tagColors[typeOption],
                                    }}
                                    onClick={() => {
                                        onSetTagType(
                                            typeOption.toUpperCase(),
                                            tagId,
                                            itemId
                                        );
                                        setDropdownIsOpen(false);
                                    }}
                                >
                                    <FontAwesome name={icons[typeOption]} />
                                </button>
                            );
                        })}
                </div>
            )}
        </div>
    );
};

export default TagDropdown;
