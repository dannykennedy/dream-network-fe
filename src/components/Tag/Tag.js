import React, { useState } from "react";
import FontAwesome from "react-fontawesome";
import AutosizeInput from "react-input-autosize";
import Spinner from "../Spinner";
import TagDropdown from "./TagDropdown";
import { connect } from "react-redux";
import {
    editTagInCurrentlyEditingPost as _editTagInCurrentlyEditingPost,
    markTagAsDeletedInCurrentlyEditingPost as _markTagAsDeletedInCurrentlyEditingPost,
} from "../../ducks/posts";
import "./Tag.css";
import { icons } from "../../theme/icons";
import { mapEntitiesToTypes } from "./tagTypes";

export function Tag({
    name,
    type,
    tagId,
    noteId,
    user,
    editing,
    editTagInCurrentlyEditingPost,
    markTagAsDeletedInCurrentlyEditingPost,
}) {
    type = mapEntitiesToTypes[type];
    console.log("mapEntitiesToTypes[type]", type);

    const [tagName, setTagName] = useState(name);

    return (
        <div className={"tag " + type} id={tagId}>
            <div className="tag-wrapper">
                {!editing && icons[type] && icons[type] !== "tag" && (
                    <div className="tag-icon">
                        <FontAwesome name={icons[type]} />
                    </div>
                )}
                {!icons[type] && (
                    <div className="tag-icon">
                        <Spinner height={"16px"} />
                    </div>
                )}
                {editing && icons[type] && (
                    <TagDropdown tagType={type} tagId={tagId} noteId={noteId} />
                )}
                {editing ? (
                    <AutosizeInput
                        name="form-field-name"
                        className="tag-edit-input"
                        value={tagName}
                        onChange={function(event) {
                            // Change local state
                            setTagName(event.target.value);
                            // Change global state
                            editTagInCurrentlyEditingPost(
                                tagId,
                                event.target.value,
                                noteId
                            );
                        }}
                    />
                ) : (
                    <div className="tag-name">
                        <span>{name}</span>
                    </div>
                )}
                {user && editing && (
                    <div className="tag-options">
                        <div
                            className="tag-remove-icon tag-icon"
                            onClick={function() {
                                markTagAsDeletedInCurrentlyEditingPost(
                                    tagId,
                                    noteId
                                );
                            }}
                        >
                            <FontAwesome name="times" />
                        </div>
                    </div>
                )}
            </div>
        </div>
    );
}

// these parts of state are passed in as props
const mapStateToProps = state => {
    return {
        user: state.user,
    };
};

const mapDispatchToProps = {
    markTagAsDeletedInCurrentlyEditingPost: _markTagAsDeletedInCurrentlyEditingPost,
    editTagInCurrentlyEditingPost: _editTagInCurrentlyEditingPost,
};

export default connect(mapStateToProps, mapDispatchToProps)(Tag);
