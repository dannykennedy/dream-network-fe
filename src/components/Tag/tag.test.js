import renderer from "react-test-renderer";
import React from "react";
import { Tag } from "./Tag";

const numItems = 11;

test("Number of items = 11", () => {
    expect(numItems).toBe(11);
});

const arr = ["dog", "cat"];

test("Array contains a cat", () => {
    expect(["dog", "cat"]).toEqual(expect.arrayContaining(arr));
});

// Objects
const obj = { title: "Dan", age: 34 };

test("The title is Dan", () => {
    expect(obj).toHaveProperty("title");
});

test("Dan's age is 34", () => {
    expect(obj).toHaveProperty("age", 34);
});

test("Tag snapshot test", () => {
    const component = renderer.create(
        <Tag
            name="dog"
            type="OTHER"
            tagId="5"
            noteId="6"
            user="1"
            editing={true}
            editTagInCurrentlyEditingPost={function() {
                console.log("hi");
            }}
            markTagAsDeletedInCurrentlyEditingPost={function() {
                console.log("hi");
            }}
        />
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
});
