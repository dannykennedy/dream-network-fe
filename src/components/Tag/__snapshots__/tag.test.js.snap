// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tag snapshot test 1`] = `
<div
  className="tag other"
  id="5"
>
  <div
    className="tag-wrapper"
  >
    <div
      className="tag-icon"
    >
      <span
        aria-hidden={true}
        className="fa fa-tag"
      />
    </div>
    <div
      className="tag-edit-input"
      style={
        Object {
          "display": "inline-block",
        }
      }
    >
      <input
        name="form-field-name"
        onChange={[Function]}
        style={
          Object {
            "boxSizing": "content-box",
            "width": "1px",
          }
        }
        value="dog"
      />
      <div
        style={
          Object {
            "height": 0,
            "left": 0,
            "overflow": "scroll",
            "position": "absolute",
            "top": 0,
            "visibility": "hidden",
            "whiteSpace": "pre",
          }
        }
      >
        dog
      </div>
    </div>
    <div
      className="tag-options"
    >
      <div
        className="tag-remove-icon tag-icon"
        onClick={[Function]}
      >
        <span
          aria-hidden={true}
          className="fa fa-times"
        />
      </div>
    </div>
  </div>
</div>
`;
